Sources/XcodeGenKit/PBXProjGenerator.swift

```diff
@@ -219,8 +242,72 @@ public class PBXProjGenerator {
-        try project.aggregateTargets.forEach(generateAggregateTarget)
+        let generateTargetsGroup = DispatchGroup()
+        var targetErrors: [Error] = []
+        var aggregateTargetsErrors: [Error] = []
+
+        var generatedTargetNames: [String] = []
+        var targetsPendingDependencyGeneration: [Target] = project.targets
+        while generatedTargetNames.count != project.targets.count {
+            var targetsReadyToGenerate = targetsPendingDependencyGeneration.filter({ [project] target in
+                let dependencies = target.getAllDependencies(usingProject: project)
+                return dependencies.reduce(into: true) { (result: inout Bool, dependency) in
+                    if dependency.reference.contains("/") {
+                        return // skip such dependencies because they are Project references most likely
+                    }
+                    result = result && generatedTargetNames.contains(dependency.reference)
+                }
+            })
+            if targetsReadyToGenerate.count == 0 {
+                // this means, that all pending targets depend on each other
+                // in such cases, pick atleast one target and proceed, to prevent an infinite loop
+                targetsReadyToGenerate = [targetsPendingDependencyGeneration.removeFirst()]
+            }
+            generateTargetsGroup.enter()
+            DispatchQueue.global(qos: .userInitiated).async { [weak self] in
+                targetsReadyToGenerate.forEach { target in
+                    generateTargetsGroup.enter()
+                    DispatchQueue.global(qos: .userInitiated).async {
+                        do {
+                            try self?.generateTarget(target)
+                        } catch let error {
+                            targetErrors.append(error)
+                        }
+                        generatedTargetNames.append(target.name)
+                        generateTargetsGroup.leave()
+                    }
+                }
+                generateTargetsGroup.leave()
+            }
+            generateTargetsGroup.wait()
+            targetsPendingDependencyGeneration = targetsPendingDependencyGeneration.filter({ target in
+                !generatedTargetNames.contains(target.name)
+            })
+        }
+
+        generateTargetsGroup.enter()
+        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
+            self?.project.aggregateTargets.forEach { target in
+                generateTargetsGroup.enter()
+                    do {
+                        try self?.generateAggregateTarget(target)
+                    } catch let error {
+                        aggregateTargetsErrors.append(error)
+                    }
+                generateTargetsGroup.leave()
+            }
+            generateTargetsGroup.leave()
+        }
+
+        generateTargetsGroup.wait()
+
+        guard targetErrors.isEmpty else {
```

Comment:
Because I couldn't find a way to make the `DispatchQueue.global` triggers follow a `try` pattern, I had to fall back to this, where i keep a track of all errors and then throw them once the completion block gets call